### PR TITLE
Added readBison()

### LINKED ISSUE #29: [future request] Parser: Add import method. (load result of dump method).
Hey, i`m working to implement some templating with php expression support. As for now i have to rewrite parts of zend_language_parser.y to php method calls. This can be match easer to load it from bison/yakk format (without C calls for sure) Like one i got as result of dump() method call:

```
%token T_LNUMBER T_DNUMBER T_STRING T_ARRAY T_OBJECT_OPERATOR T_WHITESPACE T_VARIABLE STR_D STR_S '?' ':' '|' '^' '&' '<' '>' '+' '-' '*' '/' '%' '(' ')' '[' ']' '$' '=' '.' ',' '~'
%left T_LOGICAL_OR
%left T_LOGICAL_AND
%precedence T_DOUBLE_ARROW
%left T_BOOLEAN_OR
%left T_BOOLEAN_AND
%nonassoc T_IS_EQUAL T_IS_NOT_EQUAL T_IS_IDENTICAL T_IS_NOT_IDENTICAL T_SPACESHIP
%nonassoc T_IS_SMALLER_OR_EQUAL T_IS_GREATER_OR_EQUAL
%left T_SL T_SR
%precedence '!'
%right T_POW
%left T_LOGICAL_XOR
%%

start: PARAMS;

PARAMS: PARAMS T_WHITESPACE PARAM
      | PARAM;

PARAM: T_STRING '=' expr
     | expr;

expr: NUMBER
    | STRING
    | VAR
    | ARRAY
    | expr T_BOOLEAN_OR expr
    | expr T_BOOLEAN_AND expr
    | expr T_LOGICAL_OR expr
    | expr T_LOGICAL_AND expr
    | expr T_LOGICAL_XOR expr
...
```

### PR DESCRIPTION
**Pull request by BenHanson**
Fixes: #29

### CODE REVIEW

File: tests/readBison.phpt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+<?php if (!extension_loaded("parle")) print "skip"; ?>
+--FILE--
+<?php 
+use Parle\Parser;
+use Parle\Lexer;
+
+$p = new Parser;
+
+$p->readBison("%%\n;start: 'a';%%\n");
+$p->build();
```
Comment (weltling) [inline tests/readBison.phpt]: Would it be expected, that any rules to be possibly added after `readBison()` call? If not, perhaps API wise it would be cleaner to have a static method that would look like 

```php
// Create and return the full object which is already in the finalized state
$p = Parser::fromBison(...);
``` 

Even if it should be possible to add more rules after reading the dump, perhaps semantically still it would make sense to have it in the static variant to save the line with the `new` call?

Thanks

File: tests/readBison.phpt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+$lex = new Lexer;
+$lex->push("a", $p->tokenId("'a'"));
+$lex->build();
+
+echo $p->validate("a", $lex) . "\n";
```
Comment (weltling) [inline tests/readBison.phpt]: Could you switch to using `var_dump` instead? It'll show the type of the returned value as well.

Thanks

File: tests/readBison.phpt [hunk 3]
```diff
@@ -0,0 +1,24 @@
+--TEST--
+readBison() test
```
Comment (weltling) [inline tests/readBison.phpt:2]: Could we have yet a test where an exception is thrown? Like perhaps feed some random garbage, etc. or something that would certainly fail it?

Thanks

**Review by weltling — CHANGES_REQUESTED** [review]
I don't see any obvious issues in the parsertl code. Could you please check the other comments, though?

Especially the approach with regard to the static vs. object method is IMHO something that should be looked at more closely. I could further tall for static for example, some factory like API could be initiated. For instance there could be another method to return a parser object like `Parser::fromBisonFile("/path/to/file")`.

Thanks

**Comment by weltling** [comment]
Ben, could you please also add like `Fixes: #29` to the description, so it creates a link between the issue and the pr. thanks

File: tests/readBison.phpt (hunk 1 above)
Comment (BenHanson) [reply tests/readBison.phpt]: Right now, it is simply an alternative to calling all the different functions on a rule object.

File: tests/readBison.phpt (hunk 2 above)
Comment (BenHanson) [reply tests/readBison.phpt]: The test now does a dump() instead and the test also supports UTF-32

File: parle.cpp [hunk 4]
```diff
@@ -1237,9 +1239,27 @@ _parser_dump(INTERNAL_FUNCTION_PARAMETERS, zend_class_entry *ce) noexcept
...
+		php_write((void*)u8.c_str(), u8.size());
 #else
-		parsertl::debug::dump(par.rules, std::cout);
+		std::stringstream ss;
+		std::string str;
+
+		parsertl::debug::dump(par.rules, ss);
+		str = ss.str();
+		php_write((void*)str.c_str(), str.size());
 #endif
```
Comment (weltling) [inline parle.cpp:1263]: This seems to be fixing #27, not related to the `readBison` feature. Should it be separated into another PR please? 

Thanks

Comment (BenHanson) [reply parle.cpp:1263]: This fix is necessary in order for the `readBison()` test to work.

Thanks

Comment (weltling) [reply parle.cpp:1263]: I see. Perhaps it just should have been approached in reverse order - first the dump fix, then the bison fix. Then one could split features by PR. Looks like it'd be best to land both in this PR then, to avoid additional inconvenience. Probably should land this one once both features are finalized then. I'll check to add more tests also once this is in.

Thanks

File: lib/lexertl14/include/lexertl/stream_num.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	{
+		ss_ << num_;
+	}
+
+	// MSVC doesn't support streaming integers etc to
+	// std::basic_ostringstream<char32_t>.
+	template<typename T>
+	void stream_num(const T num_, std::basic_ostream<char32_t>& ss_)
+	{
+		std::stringstream css_;
```
Comment (weltling) [inline lib/lexertl14/include/lexertl/stream_num.hpp:30]: I guess the build complains about missing `sstream` include here?

Thanks

Comment (BenHanson) [reply lib/lexertl14/include/lexertl/stream_num.hpp:30]: Thanks for the heads up. Fixed now.

File: parle.cpp (hunk 4 above)
Comment (BenHanson) [reply parle.cpp:1263]: Can we merge this one now?

Like you said, we can add extras later.

I'd like to address the other issues ASAP now. I'm hoping they will all be quick from now on.

I'd like to address the other std::cout calls next.

Comment (weltling) [reply parle.cpp:1263]: Yeah, it looks fine now. I think the last small piece is to add the missing header to the `package.xml` to make all green, then can continue on the fresh state.

Thanks

**Review by weltling — APPROVED** [review]
